[Merged by Bors] - feat(archive/imo/imo2019_q2): IMO 2019 Q2

jsm28 opened the pull request:
Add an IMO geometry problem to the mathlib archive.

The module doc string discusses the conventions followed to translate the informal statement to a formal one (which should probably end up somewhere more generally discussing conventions for formal statements of olympiad problems, but it seems reasonable to have it here for now).

This formalization uses a structure describing a configuration satisfying the conditions of the problem, to build up information gradually through many lemmas without needing to pass lots of hypotheses around (and to handle explicitly the symmetry in the configuration of the problem), which seems likely to be a convenient way to handle many olympiad geometry solutions.

---
<!-- The text above the `---` will become the commit message when your
PR is merged. Please leave a blank newline before the `---`, otherwise
GitHub will format the text above it as a title.

To indicate co-authors, include lines at the bottom of the commit message 
(that is, before the `---`) using the following format:

Co-authored-by: Author Name <author@email.com>

Any other comments you want to keep out of the PR commit should go
below the `---`, and placed outside this HTML comment, or else they
will be invisible to reviewers.

If this PR depends on other PRs, please list them below this comment,
using the following format:
- [ ] depends on: #abc [optional extra text]
- [ ] depends on: #xyz [optional extra text]
-->

- [x] depends on: #17992
- [x] depends on: #18069

[![Open in Gitpod](https://gitpod.io/button/open-in-gitpod.svg)](https://gitpod.io/from-referrer/)

jsm28 committed:
feat(geometry/euclidean/angle/sphere): law of sines, converse of inscribed angles

Add some (oriented angle) versions of the law of sines / sine rule,
and of the converse of angles in the same segment / opposite angles of
a cyclic quadrilateral.

As with previous lemmas in this area, more versions would be
appropriate to add in future, including unoriented angle versions and
versions of the law of sines that don't involve the circumradius and
so can be applied even in the degenerate case of collinear points, but
this seems a reasonable starting point.

jsm28 committed:
Fix some lemma names

jsm28 committed:
feat(archive/imo/imo2019_q2): IMO 2019 Q2

Add an IMO geometry problem to the mathlib archive.

The module doc string discusses the conventions followed to translate
the informal statement to a formal one (which should probably end up
somewhere more generally discussing conventions for formal statements
of olympiad problems, but it seems reasonable to have it here for
now).

This formalization uses a structure describing a configuration
satisfying the conditions of the problem, to build up information
gradually through many lemmas without needing to pass lots of
hypotheses around (and to handle explicitly the symmetry in the
configuration of the problem), which seems likely to be a convenient
way to handle many olympiad geometry solutions.

jsm28 committed:
Merge remote-tracking branch 'origin/master' into jsm28/oangle_law_sin

jsm28 committed:
Enter `euclidean_geometry.sphere` namespace

jsm28 committed:
Merge branch 'jsm28/oangle_law_sin' into jsm28/imo2019_q2

jsm28 committed:
Merge remote-tracking branch 'origin/master' into jsm28/imo2019_q2

kbuzzard commented on archive/imo/imo2019_q2.lean:
```diff
@@ -0,0 +1,576 @@
+-- A hypothesis implicit in the first named angle.
+(C_ne_P₁ : C ≠ P₁)
+(sbtw_Q_A₁_Q₁ : sbtw ℝ Q A₁ Q₁)
+(angle_CQ₁Q_eq_angle_CBA : ∠ C Q₁ Q = ∠ C B A)
```
Why do you make life harder for yourself by not using `Q Q₁ C` and `A B C` here?? Just because the C's are "inconsistently ordered" in the original equation doesn't mean you have to copy this for your configuration structure, you can just use the ordering of the question at the very end.

jsm28 replied:
The angles are consistently ordered in the original problem statement, where consistent means that, although the statement uses unoriented angles, all angles mentioned in the statement have the same orientation in the diagram used in the referenced solution (this is the normal convention for IMO problem statements nowadays). Reversing some of the angles without reversing all of them would lose that consistency, and I don't think avoiding a few uses of `angle_comm` is a significant gain.

kbuzzard replied:
Aah I see!

kbuzzard commented on archive/imo/imo2019_q2.lean:
```diff
@@ -0,0 +1,576 @@
+  P_ne_Q := cfg.P_ne_Q.symm,
+  sbtw_P_B₁_P₁ := cfg.sbtw_Q_A₁_Q₁,
+  angle_PP₁C_eq_angle_BAC :=
+    angle_comm cfg.C cfg.Q₁ cfg.Q ▸ angle_comm cfg.C cfg.B cfg.A ▸ cfg.angle_CQ₁Q_eq_angle_CBA,
```
This and 123 would have been easier with my suggested convention above (not that it matters much)

kbuzzard commented on archive/imo/imo2019_q2.lean:
```diff
@@ -0,0 +1,576 @@
+  have hsbtw := cfg.sbtw_Q_A₁_Q₁,
+  rw h at hsbtw,
+  have ha := hsbtw.angle₂₃₁_eq_zero,
+  rw [angle_CQ₁Q_eq_angle_CBA, angle_eq_zero_iff_ne_and_wbtw] at ha,
```
After `rw [angle_CQ₁Q_eq_angle_CBA] at ha` you have a hypothesis `ha: ∠ cfg.C cfg.B cfg.A = 0` and if this doesn't give you a one-line contradiction using `affine_independent_ABC : affine_independent ℝ ![A, B, C]` or `cfg.not_collinear_ABC` then probably you're missing API.

jsm28 replied:
I've filed #18069 with some such lemmas and used them to simplify this and another proof, and marked this PR as depending on #18069.

kbuzzard commented on archive/imo/imo2019_q2.lean:
```diff
@@ -0,0 +1,576 @@
+
+variables [module.oriented ℝ V (fin 2)]
+
+lemma oangle_CQ₁Q_sign_eq_oangle_CBA_sign :
```
One interpretation of the original problem is that `∡ cfg.C cfg.Q₁ cfg.Q = ∡ cfg.C cfg.B cfg.A`. Is this not how you read the question? Are angles by default not oriented? I guess they're not.

jsm28 replied:
I think the correct interpretation of the informal statement is in terms of unoriented angles, and angles should not be considered oriented by default. (And certainly should not be considered oriented mod pi by default, even though that's often convenient in proofs, because you can definitely talk about `sin` and `cos` of informal angles, and you can't do that mod pi.)

kbuzzard commented on archive/imo/imo2019_q2.lean:
```diff
@@ -0,0 +1,576 @@
+
+lemma symm_triangle_QPA₂ : cfg.symm.triangle_QPA₂ = cfg.triangle_PQB₂ :=
+by { simp_rw [triangle_PQB₂, ←symm_A₂], ext i, fin_cases i; refl }
+
```
Is it worth somehow flagging in this proof that the previous few hundred lines have been basically proving that a bunch of points in the diagram are distinct, something which would presumably typically not be mentioned at all by an IMO student? For example a comment on line 125 or so saying "the next 300 or so lines are proving things such as `Q₁ ≠ A₂` which would be typically not mentioned in a paper proof"? Or perhaps indicating something like "the proof proper starts here" because this is in some sense where the ideas start.

jsm28 replied:
I added a remark to the module doc string about how it's mostly proving nondegeneracy conditions.

eric-wieser replied:
It would be great to add some `/-! ### Subsection -/` headings to make this even clearer.

jsm28 replied:
I've added a series of such headings.

kbuzzard commented on archive/imo/imo2019_q2.lean:
```diff
@@ -0,0 +1,576 @@
+
+variables [module.oriented ℝ V (fin 2)]
+
+lemma two_zsmul_oangle_QPA₂_eq_two_zsmul_oangle_QB₂A₂ :
```
I think Oliver has brought this up in the past, but I guess it will confuse readers when the model solution says "we show angle QPA_2 equals angle QB_2A_2" and the Lean solution has this "strange" multiplication by two because we don't have oriented angles mod pi.

jsm28 replied:
I also remarked on this in the module doc string (referencing the `geometry.euclidean.angle.oriented.basic` doc string that also discusses this issue).

jsm28 replied:
Another possibility for angles mod pi, without adding a separate type for them and all the associated API for operating on that type and moving between the two types, would be to define `real.angle.eq_mod_pi` and API for that. But my guess is that wouldn't be very convenient to use at present, because I don't think we have any tactic to rewrite using such an equivalence relation (use an `eq_mod_pi` hypothesis to rewrite expressions contained within the expressions on either side of another `eq_mod_pi`), whereas we have lots of ways to rewrite with equality.

ocfnash replied:
Indeed, I speculated about introducing a type for angles mod `π` [here](https://github.com/leanprover-community/mathlib/pull/17525#issuecomment-1330799041) but I tend to agree with @jsm28 that it's probably not worth the trouble.

It's a bit disappointing we don't have a really neat solution for this, but I certainly don't think such matters should delay this PR.

kbuzzard reviewed (approved):
This is a monumental effort. Having now read through all of it, I do wonder whether it's worth flagging in the module docstring something like "the actual proof is 150 lines but it's preceded by 300 lines of boilerplate proving random stuff like B₂ ≠ Q which turns out to be needed for the proof and which humans would never write".

Whilst I am happy with the conventions for formalising IMO geometry problems, I wonder whether the module docstring of this particular IMO geometry problem is the best place to lay them out.

jsm28 committed:
feat(geometry/euclidean/angle/unoriented/affine): more collinearity lemmas

These lemmas relating collinearity and angles of zero or pi are less
general than the existing
`collinear_iff_eq_or_eq_or_angle_eq_zero_or_angle_eq_pi` from which
they are deduced, but may be more convenient to use.  Suggested by
review of #17993.

jsm28 committed:
Merge branch 'jsm28/collinear_angle_zero_pi' into jsm28/imo2019_q2

jsm28 committed:
Simplify two proofs using new collinearity lemmas

jsm28 committed:
Remark on nondegeneracy condition / oriented angles in module doc string

jsm28 committed:
Merge remote-tracking branch 'origin/master' into jsm28/imo2019_q2

eric-wieser commented on archive/imo/imo2019_q2.lean:
```diff
@@ -0,0 +1,572 @@
+-/
+import geometry.euclidean.angle.sphere
+
+/-! IMO 2019 Q2
```
```suggestion
/-!
# IMO 2019 Q2
```

eric-wieser commented on archive/imo/imo2019_q2.lean:
```diff
@@ -0,0 +1,572 @@
+
+lemma wbtw_B_A₁_C : wbtw ℝ cfg.B cfg.A₁ cfg.C := cfg.A₁_mem_BC
+
+lemma wbtw_B_Q_B₁ : wbtw ℝ cfg.B cfg.Q cfg.B₁ := cfg.Q_mem_BB₁
```
Is there a reason you don't just use this spelling directly in the fields?

jsm28 replied:
The idea would have been to express "lies on side" or "on segment" more literally. I've now changed the conventions to say such things are expressed with `wbtw` and used `wbtw` in the statement and the structure.

eric-wieser commented on archive/imo/imo2019_q2.lean:
```diff
@@ -0,0 +1,572 @@
+variables [module.oriented ℝ V (fin 2)]
+
+lemma two_zsmul_oangle_QPA₂_eq_two_zsmul_oangle_BAA₂ :
+  (2 : ℤ) • ∡ cfg.Q cfg.P cfg.A₂ = (2 : ℤ) • ∡ cfg.B cfg.A cfg.A₂ :=
```
This `(2 : ℤ) • x = (2 : ℤ) • y` statement seems to come up all over the place, yet I never remember seeing it on paper.
Is it worth giving this notion its own name?

jsm28 replied:
See the above remark (in a conversation Kevin marked resolved and I've now unresolved) about the possibility of defining `real.angle.eq_mod_pi` so we could say `x ≡[π] y` or similar instead. I expect that would be straightforward to add, but also (as noted above) harder to rewrite with because special-purpose rewrite lemmas would be needed rather than just rewriting directly with an `eq_mod_pi` hypothesis. (Though I suppose other relations in mathlib such as `modeq` already have that "harder to rewrite with" property.)

jsm28 committed:
Merge remote-tracking branch 'origin/master' into jsm28/imo2019_q2

jsm28 committed:
Update archive/imo/imo2019_q2.lean

Co-authored-by: Eric Wieser <wieser.eric@gmail.com>

jsm28 committed:
Directly use `wbtw` not `affine_segment` in hypotheses

jsm28 committed:
Add subsection headings

eric-wieser commented on archive/imo/imo2019_q2.lean:
```diff
@@ -0,0 +1,588 @@
+angle chase / concyclicity argument, where an informal solution doesn't discuss those conditions
+at all. Also note that (as described in `geometry.euclidean.angle.oriented.basic`, the oriented
+angles used are modulo `2 * π`, so parts of the angle chase that are only valid for angles modulo
+`π` (as used in the informal solution) are represented as equalities of twice angles.
```
```suggestion
at all. Also note that (as described in `geometry.euclidean.angle.oriented.basic`) the oriented
angles used are modulo `2 * π`, so parts of the angle chase that are only valid for angles modulo
`π` (as used in the informal solution) are represented as equalities of twice angles, which we write
as `(2 : ℤ) • ∡ _ _ _ = (2 : ℤ) • _ _ _`.
```

jsm28 committed:
Update archive/imo/imo2019_q2.lean

Co-authored-by: Eric Wieser <wieser.eric@gmail.com>

YaelDillies commented on archive/imo/imo2019_q2.lean:
```diff
@@ -0,0 +1,589 @@
+
+Prove that points `P`, `Q`, `P₁`, and `Q₁` are concyclic.
+
+We apply the following conventions for formalizing IMO geometry problems. A problem is assumed
+to take place in the plane unless that is clearly not intended, so it is not required to prove
+that the points are coplanar (whether or not that in fact follows from the other conditions).
+Angles in problem statements are taken to be unoriented. A reference to an angle `∠XYZ` is taken
+to imply that `X` and `Z` are not equal to `Y`, since choices of junk values play no role in
+informal mathematics, and those implications are included as hypotheses for the problem whether
+or not they follow from the other hypotheses. Similar, a reference to `XY` as a line is taken to
+imply that `X` does not equal `Y` and that is included as a hypothesis, and a reference to `XY`
+being parallel to something is considered a reference to it as a line. However, such an implicit
+hypothesis about two points being different is included only once for any given two points (even
+if it follows from more than one reference to a line or an angle), if `X ≠ Y` is included then
+`Y ≠ X` is not included separately, and such hypotheses are not included in the case where there
+is also a reference in the problem to a triangle including those two points, or to strict
+betweenness of three points including those two. If betweenness is stated, it is taken to be
+strict betweenness. However, segments and sides are taken to include their endpoints (unless
+this makes a problem false), although those degenerate cases might not necessarily have been
+considered when the problem was formulated and contestants might not have been expected to deal
+with them. A reference to a point being on a side or a segment is expressed directly with `wbtw`
+rather than more literally with `affine_segment`.
```
Could you make this a library note? It certainly is more applicable than this single problem (although I know getting a second one ready will take ages 😬)

jsm28 replied:
Now marked as a library note.

jsm28 committed:
Mark formalization conventions as a library note
